File: Year2/TDA417-DataStructures_Algorithms/Labs/Lab2/work/python/builtin_sort.py

from functools import cmp_to_key

from suffix_array import SuffixArray
from suffix_sorter import SuffixSorter


class BuiltinSort(SuffixSorter):
    def sort(self, sa: SuffixArray):
        sa.index.sort(
            key = cmp_to_key(sa.compareSuffixes)
        )

File: Year2/TDA417-DataStructures_Algorithms/Labs/Lab2/work/java/BuiltinSort.java

import java.util.Arrays;
import java.util.Comparator;

class BuiltinSort extends SuffixSorter {

    public void sort(SuffixArray sa) {
        Comparator<Integer> compare = (a,b) -> sa.compareSuffixes(a,b);
        // Java cannot sort an int[] using a custom comparator,
        // so we have to box the elements to Integer instead. 
        // We use streams for that:
        int[] sorted = Arrays
            .stream(sa.index)       // now we have an IntStream
            .boxed()                // now it's a Stream<Integer>
            .sorted(compare)        // now it's sorted
            .mapToInt(i -> i)       // now back to an IntStream
            .toArray();             // and finally an int[]
        // And finally we copy the result back into the original array:
        System.arraycopy(sorted, 0, sa.index, 0, sa.index.length);
    }

}

